### PR TITLE
Make TFM work with the nrf91 modem

### PR DESCRIPTION
**Pull request by oyvindronningstad**
~Based on #3636~

There were two things making the build fail when the modem was enabled:
1. The TFM RAM section was not placed in the correct place relative to the modem RAM.
2. When 1. was fixed, the TFM RAM was too large to fit in the 64k before the modem RAM.

This fixes those two things and adds an overlay config file to the nrf9160/download sample so that tfm can automatically be tested with a modem sample.

### CODE REVIEW

File: subsys/partition_manager/Kconfig [hunk 1]
```diff
@@ -141,6 +141,7 @@ if BUILD_WITH_TFM
 
 config PM_PARTITION_SIZE_TFM_RAM
 	hex "Memory reserved for TFM_RAM."
+	default 0x10000 if NRF_MODEM_LIB
```
Comment (hakonfam) [inline subsys/partition_manager/Kconfig:144]: Should we create a task to update this once https://github.com/nrfconnect/sdk-nrf/pull/3792 is merged? I guess we should update this once the amount of RAM used by modem lib is reduced.

File: subsys/partition_manager/Kconfig [hunk 2]
```diff
@@ -158,6 +159,17 @@ partition=TFM_EXTRA
...
+
+config TFM_PROFILE
+	string "The build profile used for TFM Secure image."
+	depends on BUILD_WITH_TFM
+	default "profile_small"
+	help
+	  Use profile_small when the modem is enabled to get the RAM usage below
+	  64 kB, so the Modem's IPC RAM area can fit at 0x10000.
+
+endif # TFM_PROFILE
```
Comment (hakonfam) [inline subsys/partition_manager/Kconfig:172]: `endif # NRF_MODEM_LIB`

Could you add a comment here to explain that we are overriding the default value for TFM_PROFILE? The helptext does not hint to this, and it looks like a new declaration.

File: samples/nrf9160/download/boards/nrf9160dk_nrf9160ns.overlay [hunk 3]
```diff
@@ -0,0 +1,10 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
```
Comment (hakonfam) [inline samples/nrf9160/download/boards/nrf9160dk_nrf9160ns.overlay:2]: 2021 I guess

File: subsys/partition_manager/pm.yml.trustzone [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# Create a span for the RAM to be configured as SECURE by the spm.
+sram_secure:
+  span: []
+  region: sram_primary
+
+sram_nonsecure:
+  span: [sram_primary]
```
Comment (hakonfam) [inline subsys/partition_manager/pm.yml.trustzone:7]: Should we add a comment here explaining what this does?

It actually looks quite confusing spanning the same label as the region name, and from the top of my head I'm unable to see why its needed, but I guess it is :D

File: scripts/partition_manager.py [hunk 5]
```diff
@@ -312,6 +300,11 @@ def resolve(reqs, dp):
             raise PartitionError('Unable to solve the following partitions (endless loop):\n'
                                  + pformat(unsolved))
 
+    if solution[0] != "start" or solution[-1] != "end":
```
Comment (hakonfam) [inline scripts/partition_manager.py:303]: Would it be easy for you to add a test case for this? To ensure that we don't get regression.

**Review by hakonfam — CHANGES_REQUESTED** [review]
Awesome stuff!

**Comment by tejlmand** [comment]
Is this required for 1.5 ?

File: subsys/partition_manager/Kconfig (hunk 2 above)
Comment (oyvindronningstad) [reply subsys/partition_manager/Kconfig:173]: Done

File: samples/nrf9160/download/boards/nrf9160dk_nrf9160ns.overlay (hunk 3 above)
Comment (oyvindronningstad) [reply samples/nrf9160/download/boards/nrf9160dk_nrf9160ns.overlay]: Done

File: subsys/partition_manager/Kconfig (hunk 1 above)
Comment (oyvindronningstad) [reply subsys/partition_manager/Kconfig:144]: Good idea. Will do

File: subsys/partition_manager/pm.yml.trustzone (hunk 4 above)
Comment (oyvindronningstad) [reply subsys/partition_manager/pm.yml.trustzone:7]: sram_primary is the dynamic partition.

File: scripts/partition_manager.py (hunk 5 above)
Comment (oyvindronningstad) [reply scripts/partition_manager.py:303]: There are already quite a few tests for `resolve()` that use `start` and `end`. These would fail here if that didn't work. Is there a specific corner case you want to test? We can't really test-trigger this exception, since it implies a bug in the algorithm. Maybe it makes more sense to use an `assert`?

File: subsys/partition_manager/Kconfig (hunk 1 above)
Comment (oyvindronningstad) [reply subsys/partition_manager/Kconfig:144]: Done: NCSDK-8184

File: subsys/partition_manager/Kconfig [hunk 6]
```diff
@@ -158,6 +159,18 @@ partition=TFM_EXTRA
 partition-size=0x10000
 rsource "Kconfig.template.partition_size"
 
+if NRF_MODEM_LIB
+
+config TFM_PROFILE
+	string "The build profile used for TFM Secure image."
+	depends on BUILD_WITH_TFM
```
Comment (ioannisg) [inline subsys/partition_manager/Kconfig:166]: This dependence is not needed here, i guess the whole block is under BUILD_WITH_TFM

**Review by ioannisg — APPROVED** [review]
Looks fine to me. The renames of PR macros and directives seem reasonable, but need @hakonfam to do a formal review.
I left a minor comment about a Kconfig dependency

File: subsys/partition_manager/pm.yml.trustzone (hunk 4 above)
Comment (hakonfam) [reply subsys/partition_manager/pm.yml.trustzone:7]: IMO a comment explaining this would be good, since the region has the same name as the dynamic partition and this can be confusing when reading this configuration file

File: scripts/partition_manager.py (hunk 5 above)
Comment (hakonfam) [reply scripts/partition_manager.py:303]: I have misunderstood, I thought this was a bugfix (Y)

**Review by hakonfam — APPROVED** [review]

File: subsys/partition_manager/pm.yml.trustzone (hunk 4 above)
Comment (oyvindronningstad) [reply subsys/partition_manager/pm.yml.trustzone]: Done

File: scripts/requirements-ci.txt [hunk 7]
```diff
@@ -79,6 +79,7 @@ pytz==2020.5
 pyusb==1.1.1
 PyYAML==5.4.1
 recommonmark==0.6.0
+regex==2020.11.13
```
Comment (tejlmand) [inline scripts/requirements-ci.txt:82]: `regex` should also be added here:
https://github.com/nrfconnect/sdk-nrf/blob/master/scripts/requirements-build.txt

Users are not expected to install pip packages from `requirements-ci.txt` but they should still be able to run tests.
Alternatively under:
https://github.com/nrfconnect/sdk-nrf/blob/master/scripts/requirements-extra.txt

**Review by tejlmand — COMMENTED** [review]
Looks good. Minor comment, but not blocking.

File: scripts/requirements-ci.txt (hunk 7 above)
Comment (oyvindronningstad) [reply scripts/requirements-ci.txt:82]: Done

**Review by tejlmand — APPROVED** [review]
